### PR TITLE
fix: infinite loop in peer.Ancestors()

### PR DESCRIPTION
**Pull request by gaius-qi**
Signed-off-by: Gaius <gaius.qi@gmail.com>

## Description
Fix infinite loop in `peer.Ancestors()`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: scheduler/resource/peer.go [hunk 1]
```diff
@@ -387,74 +388,84 @@ func (p *Peer) Depth() int {
 			break
 		}
 
+		ancestors = append(ancestors, parent.ID)
+
 		// Prevent traversal tree from infinite loop.
-		if p.ID == parent.ID {
+		if _, found := slices.FindDuplicate(ancestors); found {
```
Comment (bigerous) [inline scheduler/resource/peer.go:394]: we can use pkg.Set instead

**Review by bigerous — COMMENTED** [review]
There are two other function which have infinite loop risk:
https://github.com/dragonflyoss/Dragonfly2/blob/bed8e7a82e8098a7ca07150114f468fbea6a1318/scheduler/resource/peer.go#L372

https://github.com/dragonflyoss/Dragonfly2/blob/bed8e7a82e8098a7ca07150114f468fbea6a1318/scheduler/resource/peer.go#L439

**Review by yxxhero — APPROVED** [review]

**Review by hyy0322 — APPROVED** [review]
lgtm

**Review by bigerous — APPROVED** [review]
LGTM

**Pull request merged**